### PR TITLE
Allow filter to be mutated and set again

### LINKED ISSUE #2486: setFilter won't update map if you update the same array that you gave the map
```js
var filter = ['in', 'thing', 'value'];
map.setFilter('layer', filter);
filter.push('otherValue');
map.setFilter('layer', filter);
```

This breaks because the map doesn't copy the filter array and then runs deepEqual on its mutated version.

### PR DESCRIPTION
**Pull request by lucaswoj**
fixes #2486

cc @tmcw @ansis

### CODE REVIEW

**Comment by ansis** [comment]
looks good to me. Are there any other objects or arrays we should be cloning when they are passed in from the style api?

**Comment by tmcw** [comment]
Filters can be nested and this only slices the top-level - if you use an operator like 'all' and modify one of the cases, I think the same bug will still be present.

**Comment by lucaswoj** [comment]
Right you are, @tmcw! Fixed in latest commit.

**Comment by lucaswoj** [comment]
@ansis I'm not 100% sure that this fault could turn into a bug but we don't presently clone style functions passed to `addLayer` or `setLayoutProperty` or `setPaintProperty`. Fixed in the next commit.

File: js/style/style_declaration.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ module.exports = StyleDeclaration;
 function StyleDeclaration(reference, value) {
     this.type = reference.type;
     this.transitionable = reference.transition;
-    this.value = value;
+    this.value = util.clone(value);
```
Comment (mourner) [inline js/style/style_declaration.js]: forgot to require `util`

**Comment by mourner** [comment]
I wonder if we should instead just drop the `deepEquals` check — this might or might not be a better performance tradeoff than doing full cloning (which may be expensive for big filters).

**Comment by lucaswoj** [comment]
Eliminating the `deepEquals` and `clone` calls would be better for performance. I worry, however, about the unexpected behaviour that would result from mutating the filter after setting it.

```js
var filter = ['in', 'class', 'highway'];
map.setFilter('roads', filter);
map.style.update(...);
filter.push('residential');
```

After these calls, `map.getFilter('roads')` would return `['in', 'class', 'highway', 'residential']` but no `residential` roads would be visible on the map or returned from `queryRenderedFeatures`

**Pull request merged**